cns/restserver/homeazcache.go

```diff
@@ -0,0 +1,166 @@
+package restserver
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/types"
+	"github.com/Azure/azure-container-networking/log"
+	"github.com/Azure/azure-container-networking/nmagent"
+	"github.com/patrickmn/go-cache"
+	"github.com/pkg/errors"
+)
+
+const (
+	GetHomeAzAPIName = "GetHomeAz"
+	ContextTimeOut   = 2 * time.Second
+	homeAzCacheKey   = "HomeAz"
+)
+
+type HomeAzCache struct {
+	nmagentClient
+	values *cache.Cache
+	// channel used as signal to end of the goroutine for populating home az cache
+	closing chan struct{}
+	// channel used as signal to block the Start() of HomeAzCache until the first request to retrieve home az completes
+	block chan struct{}
+}
+
+// NewHomeAzCache creates a new HomeAzCache object
+func NewHomeAzCache(client nmagentClient) *HomeAzCache {
+	return &HomeAzCache{
+		nmagentClient: client,
+		values:        cache.New(cache.NoExpiration, cache.NoExpiration),
+		closing:       make(chan struct{}),
+		block:         make(chan struct{}),
+	}
+}
+
+// GetHomeAz returns home az cache value directly
+func (h *HomeAzCache) GetHomeAz(_ context.Context) cns.GetHomeAzResponse {
+	return h.readCacheValue()
+}
+
+// updateCacheValue updates home az cache value
+func (h *HomeAzCache) updateCacheValue(resp cns.GetHomeAzResponse) {
+	h.values.Set(homeAzCacheKey, resp, cache.NoExpiration)
+}
+
+// readCacheValue reads home az cache value
+func (h *HomeAzCache) readCacheValue() cns.GetHomeAzResponse {
+	cachedResp, _ := h.values.Get(homeAzCacheKey)
+	return cachedResp.(cns.GetHomeAzResponse)
+}
+
+// Start starts a new thread to refresh home az cache
+func (h *HomeAzCache) Start(retryIntervalInSecs time.Duration) {
```

Comment:
I've always found it better to be clear about where goroutines are started and where they die for daemon-like things. Otherwise you lose track of them and they leak. Functions that serve as constructors / factory functions roughly transform configuration into structs that are ready to run. One advantage is that you can have a whole bunch of things that conform to:

```go
type Runner struct {
       Start(context.Context) error
}
```

Collect all of them into a `[]Runner`, and opaquely start all of the goroutines associated with each daemon-like thing. That is, of course, spoiled here because we're accepting additional configuration in `Start()`. I didn't flag it though, since it's an easy change and not a hill I wanted to die on.